### PR TITLE
Liveness

### PR DESCRIPTION
**Pull request by miekg**

### CODE REVIEW

**Comment by miekg** [comment]
@johnbelamaric this is a generic implementation that can about as a result of our discussion in https://github.com/coredns/coredns/pull/1444#issuecomment-361972297

**Comment by miekg** [comment]
I'll probably rework *forward*'s healtchecking to use this package and see if things actually become more efficient and simpler.

File: plugin/pkg/up/up_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+			atomic.AddInt32(&hits, 1)
+		}))
+	defer web.Close()
+
+	pr := New()
+	defer pr.Stop()
+	i := 0
+	wg.Add(1)
+	upfunc := func(s string) bool {
+		http.Get(s)
```
Comment (johnbelamaric) [inline plugin/pkg/up/up_test.go]: your test could be a lot simpler if it didn't actually do an HTTP Get. You wouldn't need the `web` server at all. It doesn't matter to this "up" machinery what the upfunc actually does.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Liveness (#14..." ]
>your test could be a lot simpler if it didn't actually do an HTTP Get. You wouldn't need the `web` server at all. It doesn't matter to this "up" machinery what the upfunc actually does.

true, was more a test if it could actually be used for something like a health 
check.

simplified.

Thanks!

**Comment by miekg** [comment]
All updated.

I still might revert this if I can't make *forward* use this properly or when the resulting code doesn't look all the simpler. For now I think this will make a difference, code wise and the number of HC qps that we send.

**Comment by johnbelamaric** [comment]
Did you push new commits? I still see the HTTP server in there. Not a big deal, but I thought you were going to remove that.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Liveness (#14..." ]
>Did you push new commits? I still see the HTTP server in there. Not a big deal, but I thought you were going to remove that.

git troubles, had to reclone the repo - should be gone though. I'll double check 
once I'm back home (on my laptop)

**Comment by miekg** [comment]
found the missing tests; code updated.

**Review by johnbelamaric — APPROVED** [review]

**Pull request merged**